### PR TITLE
[Bugfix] Missing cached item in the MultiModalReceiverCache

### PR DESCRIPTION
**Pull request by knlnguyen1802**
## Purpose
Fix https://github.com/vllm-project/vllm/issues/26195

As noted by @ywang96 in https://github.com/vllm-project/vllm/issues/26195#issuecomment-3367725268  the reported error is valid and arises under specific multi-image request conditions.

## Reproduce
I reproduce the error using Qwen2.5-VL-3B and send 20 requests, each request contain 20 images size 2048x2048

## Root Cause
When serving a request that contains multiple images, during the get_and_update_item loop, items corresponding to later images may be evicted from the cache.
This eviction triggers an AssertionError in the P0 process while the cache for earlier items in the same request has already been updated.

As a result, the request is not forwarded to P1, leading to cache inconsistency between P0 and P1.

## Fix

The proposed fix modifies the behavior before defining mm_is_cached to ensure that all multimodal items in a request are accessed (touched) prior to processing.

By doing this, it prevents premature eviction during the get_and_update_item loop and guarantees cache consistency across processes.

Assumption:
The total size of all multimodal (MM) items in a single request does not exceed the configured mm-processor-cache-size.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

Test case to reproduce error for LRUCache
<details>
<summary><strong>Test script</strong></summary>

<br>

```python
import pytest
from types import SimpleNamespace
import torch
from vllm.utils.cache import LRUCache
from vllm.multimodal.cache import MultiModalProcessorSenderCache, MultiModalReceiverCache
from vllm.multimodal.inputs import MultiModalKwargsItem

# ---- Mock configuration helpers ----

class MockModelConfig:
    def __init__(self, cache_gb: float):
        self._cache_gb = cache_gb

    def get_multimodal_config(self):
        # Simulate what real ModelConfig returns
        return SimpleNamespace(mm_processor_cache_gb=self._cache_gb)

def make_dummy_item(name):
    """Creates a fake multi-modal item."""
    return MultiModalKwargsItem.dummy(name)


# ---------------- Test function ----------------
# Target image_A will get evict during process of request 2

def test_multimodal_sender_receiver_cache_eviction():
    #
    # SETUP
    #
    model_config = MockModelConfig(cache_gb=1e-9)

    sender_cache = MultiModalProcessorSenderCache(model_config)
    receiver_cache = MultiModalReceiverCache(model_config)

    # Replace internal caches with small LRU for easy testing
    sender_cache._cache = LRUCache(6, getsizeof=lambda x: 1)
    receiver_cache._cache = LRUCache(6, getsizeof=lambda x: 1)

    # REQUEST 1: 4 unique image hashes
    request1_hashes = ["image_A", "image_B", "image_C", "image_D"]
    request1_items = {h: make_dummy_item(h) for h in request1_hashes}

    # REQUEST 2: 4 images, one shared ("image_A")
    request2_hashes = ["image_E", "image_F", "image_G", "image_A"]
    request2_items = {h: make_dummy_item(h) for h in request2_hashes}

    ##########################
    # STEP 1: Request 1 send
    ##########################
    sender_is_cached_item_req1 = [
        sender_cache.is_cached_item(h) for h in request1_hashes
    ]
    print("Request 1 cache hits (Sender):", sender_is_cached_item_req1)
    # Expect all False since empty

    ###########################
    # Process request 1 for P0 Cache
    ###########################
    for i, h in enumerate(request1_hashes):
        # Use precomputed cache state
        is_cached = sender_is_cached_item_req1[i]
        item_tuple = (request1_items[h], []) if not is_cached else None

        sender_cache.get_and_update_item(item_tuple, h)
    
    ###########################
    # Process request 1 for P1 Cache
    ###########################
    for h in request1_hashes:
        receiver_cache.get_and_update_item(request1_items[h], h)

    # After request 1, caches each contain image_A–D
    print("Sender contents after Req1:", list(sender_cache._cache.keys()))
    print("Receiver contents after Req1:", list(receiver_cache._cache.keys()))

    ##########################
    # STEP 2: Request 2 send
    ##########################
    sender_is_cached_item_req2 = [
        sender_cache.is_cached_item(h) for h in request2_hashes
    ]
    print("Request 2 cache hits (Sender):", sender_is_cached_item_req2)
    # image_A should be True, others False

    ###########################
    # Process request 2 for P0 Cache
    ###########################
    for i, h in enumerate(request2_hashes):
        # Use precomputed cache state again
        is_cached = sender_is_cached_item_req2[i]
        item_tuple = (request2_items[h], []) if not is_cached else None
        print(f"Request 2: key={h} | cached={is_cached}")

        sender_cache.get_and_update_item(item_tuple, h)
    
    ###########################
    # Process request 2 for P1 Cache
    ###########################
    for h in request2_hashes:
        receiver_cache.get_and_update_item(request2_items[h], h)

    ##########################
    # STEP 3: Final states
    ##########################
    print("Sender contents after Req2:", list(sender_cache._cache.keys()))
    print("Receiver contents after Req2:", list(receiver_cache._cache.keys()))

    sender_keys = list(sender_cache._cache.keys())
    receiver_keys = list(receiver_cache._cache.keys())

    print("✅ LRU logic and re-use verified successfully.")


# Run manually
if __name__ == "__main__":
    test_multimodal_sender_receiver_cache_eviction()
```
</details>

Test case to reproduce error for SHM Cache
<details>
<summary><strong>Test script</strong></summary>

<br>

```python
import pytest
import multiprocessing as mp
from multiprocessing.synchronize import Lock as LockType
from typing import Any
from unittest.mock import MagicMock, patch

import torch
from vllm.config import ModelConfig, VllmConfig, ParallelConfig
from vllm.multimodal.cache import (
    ShmObjectStoreSenderCache,
    ShmObjectStoreReceiverCache,
)
from vllm.multimodal.inputs import (
    MultiModalKwargsItem,
    MultiModalFieldElem,
    MultiModalBatchedField,
)
from vllm.utils.mem_constants import MiB_bytes


def create_mock_image_data(size_mb: int) -> torch.Tensor:
    """Create mock image tensor of specified size in MB."""
    num_elements = (size_mb * MiB_bytes) // 4  # 4 bytes per float32
    return torch.randn(num_elements, dtype=torch.float32)


def create_mock_mm_kwargs_item(modality: str = "image") -> MultiModalKwargsItem:
    """Create a mock MultiModalKwargsItem with 1MB of data."""
    image_data = create_mock_image_data(1)  # 1MB
    
    elem = MultiModalFieldElem(
        modality=modality,
        key="image",
        data=image_data,
        field=MultiModalBatchedField(),
    )
    return MultiModalKwargsItem.from_elems([elem])


def create_mock_vllm_config(cache_size_gb: float = 0.003) -> VllmConfig:
    """Create mock VllmConfig with specified cache size."""
    mock_config = MagicMock(spec=VllmConfig)
    
    # Setup parallel config
    mock_parallel_config = MagicMock(spec=ParallelConfig)
    mock_parallel_config.world_size = 1
    mock_parallel_config._api_process_count = 1
    mock_parallel_config.data_parallel_size = 1
    mock_parallel_config.data_parallel_external_lb = False
    
    # Setup model config
    mock_model_config = MagicMock(spec=ModelConfig)
    mock_mm_config = MagicMock()
    mock_mm_config.mm_processor_cache_gb = cache_size_gb  # 3MB total
    mock_mm_config.mm_shm_cache_max_object_size_mb = 2  # Max 2MB per object
    mock_model_config.get_multimodal_config.return_value = mock_mm_config
    
    mock_config.parallel_config = mock_parallel_config
    mock_config.model_config = mock_model_config
    
    return mock_config


@pytest.mark.parametrize("cache_size_gb", [0.003])  # 6MB
def test_cache_eviction_during_request_processing(cache_size_gb: float):
    """
    Test that image_A gets evicted during request2 processing after image_G is added.
    
    Scenario:
    - Cache size: 3MB
    - Each image: 1MB
    - Request1: 3 images (A, B, C) = 3MB
    - Request2: 3 images (E, F, A) = 3MB
    
    Expected eviction during Request2:
    After adding image_R (1st image in request2):
    - Cache has: A, B, C, E (4MB > 3MB limit)
    - Need to free space → Evict A,B,C (FIFO, oldest)
    - When P1 tries to get A later, it should fail
    """
    vllm_config = create_mock_vllm_config(cache_size_gb)
    
    # Create sender cache (P0)
    sender_cache = ShmObjectStoreSenderCache(vllm_config)
    
    # Create shared lock for receiver
    shared_lock = mp.Lock()
    
    # Create receiver cache (P1)
    receiver_cache = ShmObjectStoreReceiverCache(vllm_config, shared_lock)
    
    # Image hashes
    request1_hashes = ["image_A", "image_B", "image_C"]
    request2_hashes = ["image_E", "image_F", "image_A"]
    #request2_hashes = ["image_A", "image_B", "image_C"]
    
    sender_is_cached_item_req1 = [
        sender_cache.is_cached_item(h) for h in request1_hashes
    ]
    # ===== PROCESS REQUEST 1 =====
    print("\n=== PROCESSING REQUEST 1 ===")
    request1_items = []
    request1_cached_items = []
    
    for i,mm_hash in enumerate(request1_hashes):
        is_cached = sender_is_cached_item_req1[i]
        # Create image item (1MB each)
        print(f"Is cached for hash {mm_hash} is {is_cached}")
        mm_item = (create_mock_mm_kwargs_item(), []) if not is_cached else None
        print(f"mm_item for hash {mm_hash} is {mm_item}")
        # Sender: get_and_update_item
        sender_result = sender_cache.get_and_update_item(mm_item, mm_hash)
        # print(f"Sender request is {sender_result}")
        request1_items.append(sender_result)
        
        print(f"Request1 - Sender processed {mm_hash}: "
              f"cached={mm_hash in sender_cache._shm_cache.key_index}")
        print(f"Request1 during process - Sender cached keys: {list(sender_cache._shm_cache.key_index.keys())}")

    for i,mm_hash in enumerate(request1_hashes):
        # Receiver: get_and_update_item (must be called after sender)
        receiver_result = receiver_cache.get_and_update_item(request1_items[i][0], mm_hash)
    
    print(f"\nRequest1 Complete - Sender cache size: {len(sender_cache._shm_cache.key_index)} items")
    print(f"Request1 Complete - Sender cached keys: {list(sender_cache._shm_cache.key_index.keys())}")
    
    # # ===== PROCESS REQUEST 2 =====
    print("\n=== PROCESSING REQUEST 2 ===")
    request2_items = []
    request2_cached_items = []
    sender_is_cached_item_req2 = [
        sender_cache.is_cached_item(h) for h in request2_hashes
    ]
    print(f"Before process request 2 sender_is_cached_item_req2 is {sender_is_cached_item_req2}")
    for i, mm_hash in enumerate(request2_hashes):
        # Create image item (1MB each)
        is_cached = sender_is_cached_item_req2[i]
        mm_item = (create_mock_mm_kwargs_item(), []) if not is_cached else None
        print(f"Request2 before process - Sender cached keys: {list(sender_cache._shm_cache.key_index.keys())}")
        # Sender: get_and_update_item
        sender_result = sender_cache.get_and_update_item(mm_item, mm_hash)
        request2_items.append(sender_result)
        
        sender_cached_keys = list(sender_cache._shm_cache.key_index.keys())
        print(f"Request2 - Sender processed {mm_hash}: "
              f"cached={mm_hash in sender_cache._shm_cache.key_index}")
        print(f"Request2 during process - Sender cached keys: {list(sender_cache._shm_cache.key_index.keys())}")
    
    for i,mm_hash in enumerate(request2_hashes):
        # Receiver: get_and_update_item (must be called after sender)
        receiver_result = receiver_cache.get_and_update_item(request2_items[i][0], mm_hash)
    
    # Cleanup
    sender_cache.clear_cache()
    receiver_cache.clear_cache()


if __name__ == "__main__":
    # Run the test directly
    test_cache_eviction_during_request_processing(cache_size_gb=0.003)
    print("\n✓✓✓ All tests passed! ✓✓✓")
```
</details>

Co-Author: @fake0fan 

cc: @ywang96 @DarkLight1337

### CODE REVIEW

File: vllm/multimodal/processing.py [hunk 1]
```diff
@@ -1724,6 +1724,13 @@ def _merge_mm_kwargs(
             modality: cache.is_cached(hashes) for modality, hashes in mm_hashes.items()
         }
 
+        # Need to touch all mm hashes before update to avoid hash in updated
+        # list evict during update
+        updated_mm_hashes = [
+            item_hash for hashes in mm_hashes.values() for item_hash in hashes
+        ]
+        cache.update_cache_eviction_order(updated_mm_hashes)
```
Comment (DarkLight1337) [inline vllm/multimodal/processing.py]: Need to call this in P1 as well I think

File: vllm/multimodal/cache.py [hunk 2]
```diff
@@ -703,6 +769,10 @@ def get_and_update_item(
 
         return mm_item
 
+    @override
+    def update_cache_item_eviction_order(self, mm_hash: str) -> None:
```
Comment (DarkLight1337) [inline vllm/multimodal/cache.py]: I guess this also fails the test in https://github.com/vllm-project/vllm/pull/28496#discussion_r2568606519?

Comment (knlnguyen1802) [reply vllm/multimodal/cache.py]: Yes it fail the test because the update_cache_item_eviction_order is separate from the test

File: vllm/multimodal/processing.py (hunk 1 above)
Comment (knlnguyen1802) [reply vllm/multimodal/processing.py]: It is called in get_and_update_features

Comment (DarkLight1337) [reply vllm/multimodal/processing.py]: I see

File: tests/multimodal/test_cache.py [hunk 3]
```diff
@@ -225,3 +235,289 @@ def test_ipc_enable_disable_consistency(is_cached_calls_per_iter):
         vllm_config_ipc_enabled,
         is_cached_calls_per_iter=is_cached_calls_per_iter,
     )
+
+
+def _run_test_cache_eviction_shm(
```
Comment (DarkLight1337) [inline tests/multimodal/test_cache.py:355]: Why do we need a different request sequence for SHM? Is the request sequence for the LRU test not sufficient to cover some edge case?

Comment (knlnguyen1802) [reply tests/multimodal/test_cache.py:355]: The eviction of SHM is different, and it can only evict from the beginning, so it need more time to found a good test that apply for both of them

File: vllm/multimodal/cache.py [hunk 4]
```diff
@@ -501,6 +530,15 @@ def get_and_update_item(
             logger.debug("Failed to cache mm_input with hash %s: %s", mm_hash, e)
             return mm_item
 
+    @override
+    def update_cache_item_eviction_order(
```
Comment (DarkLight1337) [inline vllm/multimodal/cache.py]: To make the typing correct, we should define separate `update_cache_item_eviction_order` methods for the P0 cache (does not need `mm_item`) and P1 cache (needs `mm_item`).

File: tests/multimodal/test_cache.py (hunk 3 above)
Comment (knlnguyen1802) [reply tests/multimodal/test_cache.py:355]: I will add description on top of both case to explain what happen

File: vllm/multimodal/cache.py (hunk 4 above)
Comment (DarkLight1337) [reply vllm/multimodal/cache.py]: Also maybe we can just rename the method to `touch_cache`.

Edit: I guess strictly speaking, for SHM cache this is not really "touch", but more like temporarily pinning the item.

File: vllm/distributed/device_communicators/shm_object_storage.py [hunk 5]
```diff
@@ -626,6 +625,39 @@ def get(self, address: int, monotonic_id: int) -> Any:
 
         return obj
 
+    def touch(
+        self, key: str, address: int = 0, monotonic_id: int = 0, is_writer: bool = True
```
Comment (DarkLight1337) [inline vllm/distributed/device_communicators/shm_object_storage.py]: `is_writer` argument is redundant, we can just check `self._reader_lock`

File: tests/multimodal/test_cache.py (hunk 3 above)
Comment (DarkLight1337) [reply tests/multimodal/test_cache.py:355]: I see, ok. Can you keep the functions in the correct order then? (functions related to LRU cache should be grouped together and those for SHM cache should be grouped together separately)

File: tests/multimodal/test_cache.py [hunk 6]
```diff
@@ -225,3 +235,289 @@ def test_ipc_enable_disable_consistency(is_cached_calls_per_iter):
...
+        p1_cache.get_and_update_item(mm_item, mm_hash)
+
+    expected_hashes = ["image_A", "image_B", "image_C"]
+    assert list(p0_cache._shm_cache.key_index.keys()) == expected_hashes
+
+    ##########################
+    # STEP 2: Request 2 send
+    ##########################
+    sender_is_cached_item_req2 = p0_cache.is_cached(request2_hashes)
+    # Note that the result for image_A is False because it will be evicted,
```
Comment (DarkLight1337) [inline tests/multimodal/test_cache.py]: The comment here is incorrect

File: tests/multimodal/test_cache.py [hunk 7]
```diff
@@ -225,3 +235,289 @@ def test_ipc_enable_disable_consistency(is_cached_calls_per_iter):
...
+    elif isinstance(p0_cache, ShmObjectStoreSenderCache):
+        assert list(p0_cache._p0_cache) == expected_hashes
+    if isinstance(p1_cache, MultiModalReceiverCache):
+        assert list(p1_cache._cache.order) == expected_hashes
+
+    ##########################
+    # STEP 2: Request 2 send
+    ##########################
+    sender_is_cached_item_req2 = p0_cache.is_cached(request2_hashes)
+    # Note that the result for image_A is False because it will be evicted,
```
Comment (DarkLight1337) [inline tests/multimodal/test_cache.py]: Same here, need to update the comment

File: vllm/distributed/device_communicators/shm_object_storage.py (hunk 5 above)
Comment (knlnguyen1802) [reply vllm/distributed/device_communicators/shm_object_storage.py]: Fix it

File: vllm/multimodal/cache.py (hunk 4 above)
Comment (knlnguyen1802) [reply vllm/multimodal/cache.py]: I decide to name them differently touch_cache_item for sender and touch_cache_feature for receiver.
Consider name it touch_cache_sender and touch_cache_receiver

Comment (DarkLight1337) [reply vllm/multimodal/cache.py]: I prefer the latter as "item" is used for both caches

File: vllm/multimodal/cache.py [hunk 8]
```diff
@@ -610,11 +633,33 @@ def get_and_update_features(
...
+        """
+        for feature in mm_features:
+            self.touch_cache_receiver(feature.identifier, feature.data)
+
         for feature in mm_features:
             feature.data = self.get_and_update_item(feature.data, feature.identifier)
         return mm_features
 
+    @abstractmethod
+    def touch_cache_receiver(self, mm_hash: str, mm_item: _I | None = None) -> None:
```
Comment (DarkLight1337) [inline vllm/multimodal/cache.py]: The `mm_item` argument shouldn't be optional anymore, since it's guaranteed to be passed inside `get_and_update_features`

Comment (knlnguyen1802) [reply vllm/multimodal/cache.py]: Fix it

File: vllm/distributed/device_communicators/shm_object_storage.py (hunk 5 above)
Comment (DarkLight1337) [reply vllm/distributed/device_communicators/shm_object_storage.py]: Doesn't look resolved yet

Comment (knlnguyen1802) [reply vllm/distributed/device_communicators/shm_object_storage.py]: I see, you mean remove the param too.

Comment (DarkLight1337) [reply vllm/distributed/device_communicators/shm_object_storage.py]: Yes

Comment (knlnguyen1802) [reply vllm/distributed/device_communicators/shm_object_storage.py]: Thanks, should be fixed now

File: vllm/distributed/device_communicators/shm_object_storage.py [hunk 9]
```diff
@@ -626,6 +625,40 @@ def get(self, address: int, monotonic_id: int) -> Any:
...
+    ) -> None:
+        """
+        Touch an existing cached item to update its eviction status.
+
+        For writers (ShmObjectStoreSenderCache): Increment writer_flag
+        For readers (ShmObjectStoreReceiverCache): Increment reader_count
+
+        Args:
+            key: String key of the object to touch
+            is_writer: If True, increment writer_flag (sender side).
```
Comment (DarkLight1337) [inline vllm/distributed/device_communicators/shm_object_storage.py]: Update the docstring as well

File: vllm/multimodal/cache.py [hunk 10]
```diff
@@ -610,11 +650,37 @@ def get_and_update_features(
...
+
         for feature in mm_features:
             feature.data = self.get_and_update_item(feature.data, feature.identifier)
         return mm_features
 
+    @abstractmethod
+    def touch_cache_receiver(
+        self,
+        mm_hash: str,
+        mm_item: MultiModalKwargsItem | None = None,
```
Comment (DarkLight1337) [inline vllm/multimodal/cache.py:656]: It is ok to make this non-optional since it's always passed in `get_and_update_features`

Comment (knlnguyen1802) [reply vllm/multimodal/cache.py:656]: I think it is ok, since the mm_item either None or MultiModalKwargsItem when called in get_and_update_features

File: vllm/distributed/device_communicators/shm_object_storage.py (hunk 9 above)
Comment (knlnguyen1802) [reply vllm/distributed/device_communicators/shm_object_storage.py]: Fixed it

File: vllm/distributed/device_communicators/shm_object_storage.py [hunk 11]
```diff
@@ -626,6 +625,40 @@ def get(self, address: int, monotonic_id: int) -> Any:
...
+            # Writer side: increment writer_flag to raise eviction threshold
+            self.increment_writer_flag(monotonic_id)
+        else:
+            with (
+                self._reader_lock,
+                self.ring_buffer.access_buf(address) as (data_view, _),
+            ):
+                reader_count = self.ring_buffer.byte2int(data_view[: self.flag_bytes])
+
+                # Avoid increasing flag on newly added item (same as original note)
```
Comment (DarkLight1337) [inline vllm/distributed/device_communicators/shm_object_storage.py]: Where is this "original note"? I can't find it

File: vllm/multimodal/cache.py [hunk 12]
```diff
@@ -302,6 +302,22 @@ def is_cached(self, mm_hashes: list[str]) -> list[bool]:
         """
         return [self.is_cached_item(mm_hash) for mm_hash in mm_hashes]
 
+    @abstractmethod
+    def touch_cache_sender(
+        self,
+        mm_hash: str,
+    ) -> None:
```
Comment (DarkLight1337) [inline vllm/multimodal/cache.py]: Nit: Make this a single line

File: vllm/distributed/device_communicators/shm_object_storage.py (hunk 11 above)
Comment (knlnguyen1802) [reply vllm/distributed/device_communicators/shm_object_storage.py]: It should be to be sync with sender. Let's me fix it

File: vllm/multimodal/cache.py (hunk 12 above)
Comment (knlnguyen1802) [reply vllm/multimodal/cache.py]: Fix all of it.

File: tests/multimodal/test_cache.py (hunk 3 above)
Comment (knlnguyen1802) [reply tests/multimodal/test_cache.py:355]: Fixed it.

File: tests/multimodal/test_cache.py [hunk 13]
```diff
@@ -225,3 +235,296 @@ def test_ipc_enable_disable_consistency(is_cached_calls_per_iter):
...
+        p1_cache.get_and_update_item(request2_items[h], h)
+
+    print(list(p0_cache._cache.order))
+    expected_hashes = ["image_D", "image_E", "image_A", "image_C"]
+    if isinstance(p0_cache, MultiModalProcessorSenderCache):
+        assert list(p0_cache._cache.order) == expected_hashes
+    elif isinstance(p0_cache, ShmObjectStoreSenderCache):
+        assert list(p0_cache._p0_cache) == expected_hashes
+    if isinstance(p1_cache, MultiModalReceiverCache):
+        assert list(p1_cache._cache.order) == expected_hashes
```
Comment (DarkLight1337) [inline tests/multimodal/test_cache.py]: Simplify this. We can narrow down the type annotations in the signature

File: tests/multimodal/test_cache.py [hunk 14]
```diff
@@ -225,3 +235,296 @@ def test_ipc_enable_disable_consistency(is_cached_calls_per_iter):
...
+    ###########################
+
+    # Touch all mm hash for P1 Cache before process
+    for mm_hash in request2_hashes:
+        p1_cache.touch_cache_receiver(mm_hash)
+
+    for h in request2_hashes:
+        p1_cache.get_and_update_item(request2_items[h], h)
+
+    print(list(p0_cache._cache.order))
```
Comment (DarkLight1337) [inline tests/multimodal/test_cache.py]: Also we don't have to print this since it's checked in the next line

File: tests/multimodal/test_cache.py [hunk 15]
```diff
@@ -225,3 +235,296 @@ def test_ipc_enable_disable_consistency(is_cached_calls_per_iter):
...
+    request2_hashes = ["image_G", "image_A"]
+    request2_items = {
+        h: MultiModalKwargsItem.dummy(
+            h, nbytes=(2 if h in request1_hashes else 1) * base_item_size
+        )
+        for h in request2_hashes
+    }
+    request2_items_p0_result = []
+
+    request3_hashes = ["image_G", "image_B"]
```
Comment (DarkLight1337) [inline tests/multimodal/test_cache.py]: Maybe we need to test multiple new items as well (e.g. if we also have `image_H` and `image_I`, then `image_I` cannot be cached as the maximum space that can be evicted is 2 units, given that only `image_A` can be evicted since `image_B` needs to be pinned)

File: vllm/multimodal/cache.py (hunk 10 above)
Comment (DarkLight1337) [reply vllm/multimodal/cache.py:656]: Ah ok, thanks for the explanation

File: tests/multimodal/test_cache.py (hunk 13 above)
Comment (knlnguyen1802) [reply tests/multimodal/test_cache.py]: Right, we does not need to check the type now

File: vllm/multimodal/processing.py [hunk 16]
```diff
@@ -1715,14 +1721,17 @@ def _merge_mm_kwargs(
...
-        mm_is_cached = {
-            modality: cache.is_cached(hashes) for modality, hashes in mm_hashes.items()
-        }
+        # Need to touch all mm hashes before update to avoid hash in updated
+        # list evict during update
+        updated_mm_hashes = [
+            item_hash for hashes in mm_hashes.values() for item_hash in hashes
+        ]
+        for mm_hash in updated_mm_hashes:
+            cache.touch_cache_sender(mm_hash)
```
Comment (DarkLight1337) [inline vllm/multimodal/processing.py]:
```suggestion
        for hashes in mm_hashes.values():
            for item_hash in hashes:
                cache.touch_cache_sender(item_hash)
```
Don't iterate twice

File: vllm/multimodal/cache.py [hunk 17]
```diff
@@ -501,6 +522,13 @@ def get_and_update_item(
             logger.debug("Failed to cache mm_input with hash %s: %s", mm_hash, e)
             return mm_item
 
+    @override
+    def touch_cache_sender(self, mm_hash: str) -> None:
+        """Touch the item in shared memory cache to prevent eviction.
+        Increments writer_flag on sender side."""
+        if self._shm_cache.is_cached(mm_hash):
```
Comment (DarkLight1337) [inline vllm/multimodal/cache.py]: Maybe we can move this check for no-op inside `shm_cache` so the interface is same as `LRUCache`

File: vllm/multimodal/cache.py [hunk 18]
```diff
@@ -353,6 +366,10 @@ def get_and_update_item(
 
         return mm_item
 
+    @override
+    def touch_cache_sender(self, mm_hash: str) -> None:
```
Comment (DarkLight1337) [inline vllm/multimodal/cache.py]: Maybe rename to `touch_sender_cache_item` and `touch_receiver_cache_item` to be consistent, since the  other methods without `item` are applied to a whole batch

File: tests/multimodal/test_cache.py (hunk 15 above)
Comment (knlnguyen1802) [reply tests/multimodal/test_cache.py]: Ok I get it now.

Comment (knlnguyen1802) [reply tests/multimodal/test_cache.py]: I change the item size now because the check of shm cache is strict (If free up 2MB item then cannot add 2 items which each of them is 1MB)
Now, the first three items is 5MB, and all other item is 2MB, when free image_A and try to add 3 more images, the first 2 can be added while the last will not.

File: vllm/multimodal/cache.py (hunk 17 above)
Comment (knlnguyen1802) [reply vllm/multimodal/cache.py]: Fix this, only need to check it for sender.

File: vllm/multimodal/cache.py (hunk 18 above)
Comment (knlnguyen1802) [reply vllm/multimodal/cache.py]: Fixed

**Review by DarkLight1337 — APPROVED** [review]
Thanks, LGTM now if tests can pass
